File: smalivm/src/test/java/org/cf/smalivm/opcode/SwitchOpTest.java
package org.cf.smalivm.opcode;

import org.cf.smalivm.VMState;
import org.cf.smalivm.VMTester;
import org.cf.smalivm.type.UnknownValue;
import org.junit.Before;
import org.junit.Test;
import org.junit.experimental.runners.Enclosed;
import org.junit.runner.RunWith;

@RunWith(Enclosed.class)
public class SwitchOpTest {

    private static final String CLASS_NAME = "Lswitch_test;";

    public static class PackedSwitch {

        private static final String METHOD_NAME = "packedSwitch()V";

        private VMState initial;

        @Before
        public void setUp() {
            initial = new VMState();
        }

        @Test
        public void testKnownPredicateVisitsExpectedLabel() {
            initial.setRegisters(0, 1, "Z");
            int[] expectedAddresses = new int[] { 0, 1, 5, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }

        @Test
        public void testUnhandledPredicateVisitsNextOp() {
            initial.setRegisters(0, 100, "S");
            int[] expectedAddresses = new int[] { 0, 1, 4, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }

        @Test
        public void testUnknownPredicateVisitsAllLabelsAndNextOp() {
            initial.setRegisters(0, new UnknownValue(), "I");
            int[] expectedAddresses = new int[] { 0, 1, 4, 5, 6, 7, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }
    }

    public static class SparseSwitch {

        private static final String METHOD_NAME = "sparseSwitch()V";

        private VMState initial;

        @Before
        public void setUp() {
            initial = new VMState();
        }

        @Test
        public void testKnownPredicateVisitsExpectedLabel() {
            initial.setRegisters(0, 1, "B");
            int[] expectedAddresses = new int[] { 0, 1, 5, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }

        @Test
        public void testUnhandledPredicateVisitsNextOp() {
            initial.setRegisters(0, 100, "I");
            int[] expectedAddresses = new int[] { 0, 1, 4, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }

        @Test
        public void testUnknownPredicateVisitsAllLabelsAndNextOp() {
            initial.setRegisters(0, new UnknownValue(), "I");
            int[] expectedAddresses = new int[] { 0, 1, 4, 5, 6, 8 };
            VMTester.testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses);
        }
    }

}

File: smalivm/src/test/java/org/cf/smalivm/opcode/SwitchOpTest.kt
package org.cf.smalivm.opcode

import org.cf.smalivm.TestState
import org.cf.smalivm.Tester.testVisitation
import org.cf.smalivm.type.UnknownValue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

class SwitchOpTest {
    companion object {
        private const val CLASS_NAME = "Lswitch_test;"
    }

    @Nested
    inner class PackedSwitch {
        private lateinit var initial: TestState
        private val METHOD_NAME = "packedSwitch()V"

        @BeforeEach
        fun setUp() {
            initial = TestState()
        }

        @Test
        fun testKnownPredicateVisitsExpectedLabel() {
            initial.setRegisters(0, 1, "Z")
            val expectedAddresses = intArrayOf(0, 1, 5, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }

        @Test
        fun testUnhandledPredicateVisitsNextOp() {
            initial.setRegisters(0, 100, "S")
            val expectedAddresses = intArrayOf(0, 1, 4, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }

        @Test
        fun testUnknownPredicateVisitsAllLabelsAndNextOp() {
            initial.setRegisters(0, UnknownValue(), "I")
            val expectedAddresses = intArrayOf(0, 1, 4, 5, 6, 7, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }
    }

    @Nested
    inner class SparseSwitch {
        private lateinit var initial: TestState
        private val METHOD_NAME = "sparseSwitch()V"

        @BeforeEach
        fun setUp() {
            initial = TestState()
        }

        @Test
        fun testKnownPredicateVisitsExpectedLabel() {
            initial.setRegisters(0, 1, "B")
            val expectedAddresses = intArrayOf(0, 1, 5, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }

        @Test
        fun testUnhandledPredicateVisitsNextOp() {
            initial.setRegisters(0, 100, "I")
            val expectedAddresses = intArrayOf(0, 1, 4, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }

        @Test
        fun testUnknownPredicateVisitsAllLabelsAndNextOp() {
            initial.setRegisters(0, UnknownValue(), "I")
            val expectedAddresses = intArrayOf(0, 1, 4, 5, 6, 8)
            testVisitation(CLASS_NAME, METHOD_NAME, initial, expectedAddresses)
        }
    }
}
